### PR TITLE
Add the IODA converter for VIIRS AOD obs

### PR DESCRIPTION
**Pull request by ypwang19**
### Description:
1. Added a converter for reading/writing aerosol optical depth (AOD) data from VIIRS to IODA.
2. Added/Edited a NetCDFToIodaConverter2d class to handle observations with different channels.
3. Added a ctest. Added VIIRS AOD sample data in testdata and sample YAML in testinput.

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
@ypwang19 can you resolve the merge conflict in utils/tests/CMakeLists.txt?

**Review by CoryMartin-NOAA — DISMISSED** [review]
This looks great, thank you @ypwang19 ! Great work!

**Review by CoryMartin-NOAA — DISMISSED** [review]

**Review by CoryMartin-NOAA — DISMISSED** [review]

File: utils/obsproc/Viirsaod2Ioda.h [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        for (int j = 0; j < dimCol; j++) {
+          if (aod550[i][j] != missingValue) {
+         // Random number generation for thinning
+             float isThin = dis(gen);
+             if (isThin > thinThreshold) {
+                preqc[i][j] = static_cast<int>(qcall[i][j]);
+                obsvalue[i][j] = static_cast<float>(aod550[i][j]);
+                lat[i][j] = lat2d[i][j];
+                lon[i][j] = lon2d[i][j];
+                float obserrorValue = 0.111431 + 0.128699 * static_cast<float>(aod550[i][j]);
```
Comment (CoryMartin-NOAA) [inline utils/obsproc/Viirsaod2Ioda.h:124]: @ypwang19 can you add comments for these 3 error assignments just to say which one is dark land, bright land, and ocean?

**Review by CoryMartin-NOAA — DISMISSED** [review]

File: utils/obsproc/Viirsaod2Ioda.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                // dark land
+                float obserrorValue = 0.111431 + 0.128699 * static_cast<float>(aod550[i][j]);
+                // ocean
+                if (qcpath[i][j] % 2 == 1) {
+                    obserrorValue = 0.00784394 + 0.219923 * static_cast<float>(aod550[i][j]);
+                }
+                // bright land
+                if (qcpath[i][j] % 4 == 2) {
+                   obserrorValue = 0.0550472 + 0.299558 *  static_cast<float>(aod550[i][j]);
+                }
```
Comment (CoryMartin-NOAA) [inline utils/obsproc/Viirsaod2Ioda.h]:
```suggestion
                }
```

Comment (CoryMartin-NOAA) [reply utils/obsproc/Viirsaod2Ioda.h]: the C++ linter doesn't like this

**Review by CoryMartin-NOAA — APPROVED** [review]
Perfect, thanks @ypwang19 !! I'll merge this tomorrow unless any of the other reviewers have any changes requested before then.

File: utils/obsproc/Viirsaod2Ioda.h (hunk 2 above)
Comment (ypwang19) [reply utils/obsproc/Viirsaod2Ioda.h]: Resolved.

File: utils/obsproc/NetCDFToIodaConverter2D.h
Comment (andytangborn) [inline utils/obsproc/NetCDFToIodaConverter2D.h:1]: I take it that [NetCDFToIodaConverter2D.h] is the equivalent of the python top level code that finds all of the VIIRS obs within the time window, and passes them to the actual IODA converter.  I always found it awkward to not have that code in github.

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+#include "ioda/Group.h"
+#include "ioda/ObsGroup.h"
+
+#include "oops/mpi/mpi.h"
+#include "oops/util/DateTime.h"
+#include "oops/util/Duration.h"
+#include "oops/util/Logger.h"
+#include "oops/util/missingValues.h"
+
+namespace gdasapp {
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:20]: I'm pretty sure you could have simply modified the iodavars struct that we have under `utils.h`, that's a lot of duplicate code, some of which has been updated and possibly debugged.

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+    // Optional metadata
+    Eigen::ArrayXXf floatMetadata_;                // Optional array of float metadata
+    std::vector<std::string> floatMetadataName_;  // String descriptor of the float metadata
+    Eigen::ArrayXXf intMetadata_;                  // Optional array of integer metadata
+    std::vector<std::string> intMetadataName_;    // String descriptor of the integer metadata
+
+    // Optional global attributes
+    std::map<std::string, std::string> strGlobalAttr_;
+
+    // Constructor
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:52]: you could just add the this constructor to the original iodavars struct, or have a case for channels in the original constructor.

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+      floatMetadata_(location_, fmnames.size()),
+      floatMetadataName_(fmnames),
+      intMetadata_(location_, imnames.size()),
+      intMetadataName_(imnames)
+    {
+      oops::Log::trace() << "IodaVars2d::IodaVars2d created." << std::endl;
+    }
+
+    // Append an other instance of IodaVars
+    void append(const IodaVars2d& other) {
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:72]: same here, that doesn't seem to complicated to generalize ...

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    // Constructor: Stores the configuration as a data members
+    explicit NetCDFToIodaConverter2d(const eckit::Configuration & fullConfig,
+                                   const eckit::mpi::Comm & comm): comm_(comm),
+                                                                   fullConfig_(fullConfig) {
+      // time window info
+      std::string winbegin;
+      std::string winend;
+      fullConfig.get("window begin", winbegin);
+      fullConfig.get("window end", winend);
+      windowBegin_ = util::DateTime(winbegin);
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:118]: window info is probably not used, at least we don't and I suspect it's the same here.

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+      // ioda output file name
+      outputFilename_ = fullConfig.getString("output file");
+      oops::Log::info() << "--- Output files: " << outputFilename_ << std::endl;
+      oops::Log::trace() << "NetCDFToIodaConverter::NetCDFToIodaConverter created." << std::endl;
+    }
+
+    // Method to write out a IODA file (writter called in destructor)
+    void writeToIoda() {
+      // Extract ioda variables from the provider's files
+      int myrank  = comm_.rank();
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:137]: we have an issue with mpi, it currently only works with containers, and we haven't implemented anything for the gathering of the metadata. I suggest you remove the mpi implementation.

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+      // Get the total number of obs across pe's
+      int nobsAll(0);
+      comm_.allReduce(nobs, nobsAll, eckit::mpi::sum());
+      gdasapp::IodaVars2d iodaVarsAll(nobsAll,
+                                    iodaVars.channel_,
+                                    iodaVars.floatMetadataName_,
+                                    iodaVars.intMetadataName_);
+
+      // Gather iodaVars arrays
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:160]: that section should go.

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+        iodaPreQc.writeWithEigenRegular(iodaVarsAll.preQc_);
+      }
+    }
+
+   private:
+    // Virtual method that reads the provider's netcdf file and store the relevant
+    // info in a IodaVars struct
+    virtual gdasapp::IodaVars2d  providerToIodaVars(const std::string fileName) = 0;
+
+    // Gather for eigen array
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:259]: that looks like the exact same function that's under utils.h ... Remove it and unless you are planning to fix the mpi issues, don't make use of mpi just yet.

File: utils/obsproc/NetCDFToIodaConverter2D.h [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+      comm_.barrier();
+      comm_.gatherv(send, recv, sizes, displs, root);
+
+      if (comm_.rank() == root) {
+        obsAllPes.segment(0, recv.size()) =
+          Eigen::Map<Eigen::Array<T, Eigen::Dynamic, 1>>(recv.data(), recv.size());
+      }
+    }
+
+    // Short-cut to create type dependent VariableCreationParameters
```
Comment (guillaumevernieres) [inline utils/obsproc/NetCDFToIodaConverter2D.h:298]: how is this different from the original function?

File: utils/test/testinput/gdas_viirsaod2ioda.yaml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+provider: VIIRSAOD
```
Comment (guillaumevernieres) [inline utils/test/testinput/gdas_viirsaod2ioda.yaml:1]: You should probably add a reference test ... We found a few bugs on our side when we did this.

**Review by guillaumevernieres — COMMENTED** [review]
Sorry for the late review @ypwang19 . I know it's been merged but I thought I'd offer my 0.02 cents. See my review comments.

I don't mind the extensive copy/paste, if the idea was to quickly put together something that works, what you did looks like a great first iteration. But if you are planning to make use of this feature in the future, we need to do a bit of cleanup. Happy to chat next Thursday, there are a few low-hanging fruit "things" that could be easily implemented.

**Comment by ypwang19** [comment]
> Sorry for the late review @ypwang19 . I know it's been merged but I thought I'd offer my 0.02 cents. See my review comments.
> 
> I don't mind the extensive copy/paste, if the idea was to quickly put together something that works, what you did looks like a great first iteration. But if you are planning to make use of this feature in the future, we need to do a bit of cleanup. Happy to chat next Thursday, there are a few low-hanging fruit "things" that could be easily implemented.

Thank you for the review! I basically copied the original NetCDFToIodaConverter.h (before it was separated to util.h) and did modification on that. That's why there are duplication. I agree that cleaning-up is necessary and I will work on that.

**Pull request merged**